### ISSUE TITLE
fastText: Cannot load ../fasttext/wiki.es/wiki.es.bin due to C++ extension failed to allocate the memory

### ISSUE THREAD

**Issue by stanlee321** [salestock/fastText.py#115]
Hi, I just tried to load the wiki.bin and wiki.vac pre-trained models and I stuck in the error 
fastText: Cannot load ../fasttext/wiki.es/wiki.es.bin due to C++ extension failed to allocate the memory
 Is this part of a hardware issue? I have right now 12 GB of RAM and a Linux environment running into a docker container.
Thanks.

**Comment by pyk** [comment]
> Is this part of a hardware issue? I have right now 12 GB of RAM and a Linux environment running into a docker container.

Hmm if you have `12Gb` free memory, it's enough to load the pre-trained models. You can run this command

```
 free -mh
```

to see available memory on your linux environment.

**Comment by stanlee321** [comment]
Solve the issue, it was just the default config of docker container of 1 GB of mem. Now with 8 GB the whole model was loaded without problem.

**Comment by pyk** [comment]
@stanlee321 I will close this issue for now. Feel free to re-open the issue.

**Comment by ytac8** [comment]
I have same error when I tried to `model = fasttext.loadmodel('../data/wiki.en.bin')` . 
The error is
```
File "fasttext/fasttext.pyx", line 154, in fasttext.fasttext.load_model (fasttext/fasttext.cpp:4292)
Exception: fastText: Cannot load ../data/wiki.en.bin due to C++ extension failed to allocate the memory
```

My PC has 32 GB memory, and the result of `free -mh` is below.
```
🐶 { free -mh                                                                         
              total        used        free      shared  buff/cache   available
Mem:            31G        1.2G        484M        147M         29G         29G
Swap:           31G        3.1G         28G

```

what should I do to use fasttext?
Thanks.

**Comment by spate141** [comment]
@ytac8 This error is mainly because the `wiki.en.bin` file is trained with newer version of fastText, and as of now you can't use current fastText.py python library to load/use newer models. 

More info: #125

**Comment by ytac8** [comment]
@spate141 
Oh, I see.
I could not find this issue. Thank you very much for your information 🥇

**Comment by zachmayer** [comment]
Were could I get an old version of `wiki.en.bin`?

**Comment by ytac8** [comment]
@zachmayer 
you can get here!
https://github.com/facebookresearch/fastText/blob/master/pretrained-vectors.md

**Comment by blorgblerg** [comment]
There are some older pre-trained models here: https://github.com/Kyubyong/wordvectors

**Comment by shgidi** [comment]
What is the last version of fastText that is compatible with the python extension? where can I find a pretrained english model that fits the python extension?

**Comment by dcsan** [comment]
same question here. it seems the version of fastText everyone will be using from the facebook repo isn't compatible with this extension... #143

**Comment by kelly-tlz** [comment]
I have the same question,  whatever model I load. @spate141 @blorgblerg

**Comment by dcsan** [comment]
the answer is not to use this python library, as it seems won't be updated to newer version of fastText.

**Comment by kelly-tlz** [comment]
can you share your code with me? thanks a lot

**Comment by kelly-tlz** [comment]
can you share your code with me? thanks a lot @dcsan

**Comment by montenegrodr** [comment]
@kelly-tlz you've to use the interactive option of fasttext. You should wrap and use stdin and stdout, something like this:

```
import subprocess

cmds = ['path/to/fasttext', 'predict', 'path/to/model', '-']

interactive_model = subprocess.Popen(cmds, stdout=subprocess.PIPE, stdin=subprocess.PIPE)

# write into stdin
interactive_model.stdin.write('text input' + '\n')

# read stdout
result = interactive_model.stdout.readline()
```

**Comment by dcsan** [comment]
yep the last part `-` is key to tell fasttext to accept stdin input

`cmds = ['path/to/fasttext', 'predict', 'path/to/model', '-']`

also when you're sending text in don't forget to add a newline \n on your final input, or it will hang waiting for input.
I ended up switching to using nodeJS so if this helps, some thing like:

```js
    const count = 5
    const opts = ['predict-prob', config.FT_MODEL_PATH, '-', count]
    const fasttext = spawn(config.FT_BINARY_PATH, opts)


  // plain text line (not segmented)
  // because we're returning a promise this func is not defined as 'async'
  predict (line, count) {
    count = count || 5
    let prom = new Promise((resolve, reject) => {
      let stdout = ''
      let child = Classifier.FastText

      child.stdout.on('data', async (data) => {
        // debug.debug('stdout.data>', String(data))
        stdout += String(data)
        let predictions = await Classifier.formatResults(stdout)
        // clear out listeners for next time func is called
        child.stdout.removeAllListeners('data')
        child.stdout.removeAllListeners('error')
        resolve(predictions)
      })

      child.stderr.on('data', (data) => {
        debug.debug('ERROR stderr.data', data)
      })

      // this doesn't usually get called
      child.on('end', async (data) => {
        debug.error('FastText process quit', String(data))
        // TODO - init again
      })

      child.on('error', (err) => {
        debug.error('ERROR FastText', err)
        reject(err)
      })

      let chunked = Segmenter.segment(line)
      debug.debug('writing to fasttext', chunked)
      try {
        child.stdin.write(chunked + '\n')
      } catch (err) {
        debug.error('ERROR', err)
        // debug.error('trying to re-init classifier', err)
        // Classifier.init(AppConfig.cfg)
      }
    })
    return (prom)
  },
```

**Comment by XinCindyChen** [comment]
I tried to use the subprocess solution, but I kept getting `[Errno 32] Broken pipe` or empty results. Anyone have a solution? I'm trying to get the vector for words and sentences from the pre-trained model based on Wikipedia text.

Below is my code, I'm hoping to get the vector for "apple", but it returned broken pipe. 
On the other hand, `echo 'apple' | ./fasttext print-word-vectors wiki.simple.bin` worked on the command line. 
```
import subprocess
cmds = ['/Users/XX_XX/fastText/fasttext', 'print-word-vectors', 'wiki.simple.bin', '-']
interactive_model = subprocess.Popen(cmds, stdout=subprocess.PIPE, stdin=subprocess.PIPE)
interactive_model.stdin.write('apple' + '\n')
result = interactive_model.stdout.readline()
```

**Comment by montenegrodr** [comment]
@XinCindyChen it seems there's no interactive option for `print-word-vectors`:

```
$ ./fasttext print-sentence-vectors
usage: fasttext print-sentence-vectors <model>

  <model>      model filename
```

**Comment by dcsan** [comment]
I think i saw somewhere that method doesn't have an interactive mode ( `-`param ). did you try it with the same args just from the command line? otherwise you'd have to dump stuff to a file and read it in one chunk. however not sure what you would do if you had another list later...

for other points, make sure to put a newline after each (and the last) data you stream in. i forgot that at first...
`child.stdin.write(chunked + '\n')`

**Comment by XinCindyChen** [comment]
@montenegrodr @dcsan you are right, I removed the last '-' from the list of cmds, it worked with the following code. The result contains the vector for "apple".  But I still don't quite understand the difference between having an interactive mode or not, since both did take an input and give an output. Could you elaborate? 

```import subprocess
cmds = ['/Users/XX_XX/fastText/fasttext', 'print-word-vectors', 'wiki.simple.bin']
interactive_model = subprocess.Popen(cmds, stdout=subprocess.PIPE, stdin=subprocess.PIPE)
interactive_model.stdin.write('apple' + '\n')
result = interactive_model.stdout.readline()```

**Comment by dcsan** [comment]
interactive mode is when you run it from the command line, and you can keep feeding new words in, like an interactive REPL.
this is useful when you just want to load up fasttext once from your own app and then keep piping more data in.

the *non*interactive way means you need to reload the app every time you want to call it. which also means loading into memory all those big model files...

**Comment by XinCindyChen** [comment]
@dcsan this makes sense, thank you very much. No wonder it was taking me forever to print vectors for multiple words/sentences using the subprocess.

**Comment by MartinXPN** [comment]
There is also a gensim wrapper for fastText. Take a look at the [documentation](https://radimrehurek.com/gensim/models/wrappers/fasttext.html) here. 

If you don't want to use gensim you still can create a wrapper class like this

```python
class FastText(object):
    def __init__(self, fasttext_lib_directory, fasttext_model_path):
        cmds = [fasttext_lib_directory, 'print-word-vectors', fasttext_model_path]  # Add '-' in the end for interactive mode, yet it didn't work for me...
        self.model = subprocess.Popen(cmds, stdout=subprocess.PIPE, stdin=subprocess.PIPE, env=os.environ.copy())

        # Test the model
        print('\nTesting the model...\nPrediction for apple: ')
        item = 'apple\n'
        item = item.encode('utf-8')
        self.model.stdin.write(item)
        result = self.model.stdout.readline()
        result = result[len(item):]
        result = np.fromstring(result, dtype=np.float32, sep=' ')
        self.vector_size = len(result)
        print('Length of word-vector is:', self.vector_size)

    def __getitem__(self, item):
        assert type(item) is str
        initial_item = item
        item = item.lower().replace('/', '').replace('-', '').replace('\\', '').replace('`', '')
        if len(item) == 0 or ' ' in item:
            raise KeyError('Could not process: ' + initial_item)

        if not item.endswith('\n'):
            item += '\n'

        item = item.encode('utf-8')
        self.model.stdin.write(item)
        self.model.stdout.flush()
        result = self.model.stdout.readline()  # Read result
        result = result[len(item):]            # Take everything but the initial item
        result = np.fromstring(result, dtype=np.float32, sep=' ')

        if len(result) != self.vector_size:
            print('Could not process: ' + item)
            raise KeyError('Could not process: ' + initial_item)
        return result
```

And then use the model like

```python
model = FastText(fasttext_lib_directory='./fastText/fasttext', fasttext_model_path='./wiki.en.bin')
print(model['apple'])
print(model['banana'])
```

**Comment by dcsan** [comment]
i understood the gensim wrapper was based on an old (incompatible) version of FT and wasn't going to be updated as the company working on it no longer uses FT or gensim. might not be too difficult to update it tho but requires knowhow of the C bindings.

**Comment by MartinXPN** [comment]
I tried loading the latest wiki.en.bin and it worked with gensim

**Comment by HoiweCheng** [comment]
i got the same problem while loading the self-trainged model

**Comment by montenegrodr** [comment]
FYI: https://github.com/facebookresearch/fastText/tree/master/python

**Comment by sayak1711** [comment]
@montenegrodr @dcsan I am trying to run fastText on Windows. However I couldn't install it using python setup.py install. So I had to install it from the [unofficial python wheel](https://www.lfd.uci.edu/~gohlke/pythonlibs/#fasttext). However when I try to load my model(created on a red hat system) it says memory allocation error(on windows). However when I use bash on windows to predict labels using this model(a bin file) I successfully get results. However since I won't be manually entering input data for predicting each time I want python to do that. So I came across this thread. subprocess and os.system by default use the command prompt on Windows for running the commands given. However I can't run fasttext on command prompt as I get the error '.' is not recognized as an internal or external command. So how do I use subprocess to run the ./fasttext predict command on bash instead of command prompt?
Here is what I tried:
`import subprocess
import os

#subprocess.Popen('C:/Windows/System32/bash.exe', stdout=subprocess.PIPE, stdin=subprocess.PIPE)
#os.system('C:/Windows/System32/bash.exe')
cmds = ['C:/Windows/System32/bash.exe', './fasttext', 'predict', 'models/mymodel.bin', '-']
interactive_model = subprocess.Popen(cmds, stdout=subprocess.PIPE, stdin=subprocess.PIPE)
# write into stdin
interactive_model.stdin.write('text input' + '\n')

# read stdout
result = interactive_model.stdout.readline()
print(result)`

**Comment by grasshopper14** [comment]
subprocess doesn't display an output in the virtual environment. Is there a way to do this in virtual env?
